java/libphonenumber/src/com/google/i18n/phonenumbers/MultiFileMetadataSourceImpl.java

```diff
@@ -73,91 +67,81 @@ public MultiFileMetadataSourceImpl(MetadataLoader metadataLoader) {
 
   @Override
   public PhoneMetadata getMetadataForRegion(String regionCode) {
-    synchronized (regionToMetadataMap) {
-      if (!regionToMetadataMap.containsKey(regionCode)) {
-        // The regionCode here will be valid and won't be '001', so we don't need to worry about
-        // what to pass in for the country calling code.
-        loadMetadataFromFile(regionCode, 0);
-      }
-    }
-    return regionToMetadataMap.get(regionCode);
+    PhoneMetadata metadata = regionToMetadataMap.get(regionCode);
+    return (metadata != null) ? metadata : loadMetadataFromFile(
+        regionCode, regionToMetadataMap, filePrefix, metadataLoader);
   }
 
   @Override
   public PhoneMetadata getMetadataForNonGeographicalRegion(int countryCallingCode) {
-    synchronized (countryCodeToNonGeographicalMetadataMap) {
-      if (!countryCodeToNonGeographicalMetadataMap.containsKey(countryCallingCode)) {
-        List<String> regionCodes =
-            CountryCodeToRegionCodeMap.getCountryCodeToRegionCodeMap().get(countryCallingCode);
-        // We can assume that if the country calling code maps to the non-geo entity region code
-        // then that's the only region code it maps to.
-        if (regionCodes.size() == 1 &&
-            PhoneNumberUtil.REGION_CODE_FOR_NON_GEO_ENTITY.equals(regionCodes.get(0))) {
-          loadMetadataFromFile(PhoneNumberUtil.REGION_CODE_FOR_NON_GEO_ENTITY, countryCallingCode);
-        }
-      }
+    PhoneMetadata metadata = countryCodeToNonGeographicalMetadataMap.get(countryCallingCode);
+    if (metadata != null) {
+      return metadata;
     }
-    return countryCodeToNonGeographicalMetadataMap.get(countryCallingCode);
+    if (isNonGeographicalCountryCallingCode(countryCallingCode)) {
+      return loadMetadataFromFile(
+          countryCallingCode, countryCodeToNonGeographicalMetadataMap, filePrefix, metadataLoader);
+    }
+    return null;
```

Comment:
maybe a comment explaining why we choose to return null here (ie, because the given code was not non-geographic).